### PR TITLE
[TACHYON-599] Update allocators and evictors to work with BlockMetadataManagerView

### PR DESCRIPTION
**Pull request by ooq**
Tests are not fixed yet.

### CODE REVIEW

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 1]
```diff
@@ -89,17 +89,19 @@ public TieredBlockStore(TachyonConf tachyonConf) throws IOException {
     mTachyonConf = Preconditions.checkNotNull(tachyonConf);
     mMetaManager = BlockMetadataManager.newBlockMetadataManager(mTachyonConf);
     mLockManager = new BlockLockManager(mMetaManager);
+    BlockMetadataManagerView initManagerView =
+        new BlockMetadataManagerView(mMetaManager, new HashSet<Integer>() , new HashSet<Long>());
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: I think you can use `Collections.emptySet()`

Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: You may want to create one view for the allocator and one to the evictor just for isolation

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 2]
```diff
@@ -436,7 +440,7 @@ private BlockMetadataManagerView getUpdatedView() {
   /** This method must be guarded by WRITE lock of mEvictionLock */
   private void freeSpaceInternal(long userId, long availableBytes, BlockStoreLocation location)
       throws IOException {
-    EvictionPlan plan = mEvictor.freeSpace(availableBytes, location);
+    EvictionPlan plan = mEvictor.freeSpaceWithView(availableBytes, location, getUpdatedView());
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: I think you need to synchronize on the pin list lock in the `getUpdatedView` method.

File: servers/src/main/java/tachyon/worker/block/evictor/GreedyEvictor.java [hunk 3]
```diff
@@ -119,34 +146,35 @@ public EvictionPlan freeSpace(long availableBytes, BlockStoreLocation location)
   }
 
   // TODO: share this as a util function as it may be useful for other Evictors.
-  private boolean canEvictBlocksFromDir(StorageDir dir, long availableBytes) {
-    return dir.getAvailableBytes() + dir.getCommittedBytes() >= availableBytes;
+  private boolean canEvictBlocksFromDir(StorageDirView dirView, long availableBytes) {
+    return dirView.getAvailableBytes() + dirView.getCommittedBytes() >= availableBytes;
```
Comment (ooq) [inline servers/src/main/java/tachyon/worker/block/evictor/GreedyEvictor.java]: update `dirView.getCommittedBytes()` -> `dirView.getEvictableBytes()`

File: servers/src/main/java/tachyon/worker/block/allocator/GreedyAllocator.java [hunk 4]
```diff
@@ -30,22 +30,40 @@
...
+  private TempBlockMeta allocateBlock(long userId, long blockId, long blockSize,
       BlockStoreLocation location) throws IOException {
     if (location.equals(BlockStoreLocation.anyTier())) {
-      // When any tier is ok, loop over all storage tier and dir, and return the first dir that has
-      // sufficient available space.
-      for (StorageTier tier : mMetaManager.getTiers()) {
-        for (StorageDir dir : tier.getStorageDirs()) {
-          if (dir.getAvailableBytes() >= blockSize) {
-            return new TempBlockMeta(userId, blockId, blockSize, dir);
+      // When any tier view is ok, loop over all tier views and dir views,
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/allocator/GreedyAllocator.java]: The comment should probably still be `When any tier is ok`?

File: servers/src/main/java/tachyon/worker/block/allocator/GreedyAllocator.java [hunk 5]
```diff
@@ -30,22 +30,40 @@
...
+      // and return the first dir view that has sufficient available space.
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/allocator/GreedyAllocator.java]: Maybe change to `and return a temp block in the first dir that has ...`?

File: servers/src/main/java/tachyon/worker/block/allocator/MaxFreeAllocator.java [hunk 6]
```diff
@@ -19,66 +19,84 @@
...
-import tachyon.worker.block.meta.StorageDir;
-import tachyon.worker.block.meta.StorageTier;
+import tachyon.worker.block.meta.StorageDirView;
+import tachyon.worker.block.meta.StorageTierView;
 import tachyon.worker.block.meta.TempBlockMeta;
 
 /**
- * An allocator that allocates a block in the storage dir with most free space. It always allocates
- * to the highest tier if the requested block store location is any tier.
+ * An allocator that allocates a block in the storage dir view with most free space.
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/allocator/MaxFreeAllocator.java]: Probably want to keep this as storage dir?

File: servers/src/main/java/tachyon/worker/block/meta/StorageDir.java [hunk 7]
```diff
@@ -34,6 +34,7 @@
 
 import tachyon.Constants;
 import tachyon.StorageDirId;
+import tachyon.worker.block.BlockMetadataManager;
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/meta/StorageDir.java]: What's this import for?

File: servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java [hunk 8]
```diff
@@ -20,7 +20,9 @@
 
 import com.google.common.base.Preconditions;
 
+import tachyon.worker.block.BlockMetadataManager;
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java]: What's this import for?

**Comment by calvinjia** [comment]
+1 LGTM

**Comment by ooq** [comment]
@apc999

**Comment by apc999** [comment]
+1 LGTM

File: servers/src/main/java/tachyon/worker/block/BlockMetadataManagerView.java [hunk 9]
```diff
@@ -107,7 +114,12 @@ public boolean isBlockEvictable(long blockId) {
   public StorageTierView getTierView(int tierAlias) throws IOException {
     // TODO: can we ensure the returning tierview is same as
     // new StorageTierView(mMetadataManager.getTier(tierAlias)) ?
-    return mTierViews.get(tierAlias);
+    StorageTierView tierView = mAliasToTierViews.get(tierAlias);
+    if (null == tierView) {
+      throw new IOException("Cannot find tier view with alias: " + tierAlias);
+    } else {
+      return tierView;
+    }
```
Comment (haoyuan) [inline servers/src/main/java/tachyon/worker/block/BlockMetadataManagerView.java]: if (null == tierView) {
      throw new IOException("Cannot find tier view with alias: " + tierAlias);
}
return tierView;

**Pull request merged**